### DISCUSSION TITLE
Optional ImageDecoder

### DISCUSSION THREAD

**Discussion by hubgit** [mozilla/pdf.js#19030] [General]
Since upgrading from [v4.7.76](https://github.com/mozilla/pdf.js/releases/tag/v4.7.76) to [v4.8.69](https://github.com/mozilla/pdf.js/releases/tag/v4.8.69) we've discovered a few issues with `ImageDecoder` in Chrome (one crash on large images, which is already fixed in Chrome Canary; one faulty image rendering when a JPEG image has an unusual color profile, which I'm going to report).

Might it be possible to have an option to disable the use of `ImageDecoder`, for situations where we know there could be a problem?

**Comment by calixteman** [comment]
Would it be possible to share the pdf with the faulty image ?

That said, in considering that `ImageDecoder` is a pretty recent feature, it needs to be stabilized, so it makes sense to have such an option.
@Snuffleupagus wdyt ?

**Reply by Snuffleupagus** [reply]
Don't we already have an option for that?
https://github.com/mozilla/pdf.js/blob/13a231cd3b95fc29f31f81344ac15e4162e9efe1/src/display/api.js#L180-L182

**Reply by hubgit** [reply]
Interesting, I suspect https://issues.chromium.org/issues/374807001 is the issue with large [edit: JPEG] images that I found to have been fixed in Chrome Canary.

Maybe `JpegStream.canUseImageDecoder` needs to read `isChrome` too?

**Reply by Snuffleupagus** [reply]
Given that the existing option is only relevant for BMP, I've submitted PR #19031 to add a more general-purpose option (that can stay even after `isChrome` is removed).

**Comment by hubgit** [comment]
The separate issue with the color profile can be avoided by setting [`colorSpaceConversion: 'none'`](https://developer.mozilla.org/en-US/docs/Web/API/ImageDecoder/ImageDecoder#colorspaceconversion) in the `ImageDecoder` constructor options, but I suspect that might have unwanted side effects.

**Reply by calixteman** [reply]
> The separate issue with the color profile can be avoided by setting [`colorSpaceConversion: 'none'`](https://developer.mozilla.org/en-US/docs/Web/API/ImageDecoder/ImageDecoder#colorspaceconversion) in the `ImageDecoder` constructor options, but I suspect that might have unwanted side effects.

Could you share this specific pdf ?

**Reply by hubgit** [reply]
The original PDF was private, but I've extracted the problematic image from the PDF, copied the color profile to a new image with `exiftool` and then converted it to PDF.

In the attached zip file are:

* `good.jpg` and `good.pdf` - an unmodified example image, converted to PDF, which has no issues.
* `bad.icc` - the color profile that's causing a problem.
* `bad.jpg` and `bad.pdf` - the example image modified with the `bad.icc` color profile, converted to PDF, which exhibits the issue when opened in https://mozilla.github.io/pdf.js/web/viewer.html using Chrome.

[icc.zip](https://github.com/user-attachments/files/17723182/icc.zip)

**Reply by hubgit** [reply]
I've used this code (with the image served from a local folder) to show that the problem exists when the image is only processed by `ImageDecoder`:

```js
const res = await fetch('http://localhost:3000/bad.jpg')
const data = await res.arrayBuffer()

const decoder = new ImageDecoder({ 
  type: 'image/jpeg', 
  data,
  // colorSpaceConversion: 'none',
})
const { image } = await decoder.decode()
decoder.close()

const canvas = document.createElement("canvas");
canvas.width = image.displayWidth;
canvas.height = image.displayHeight;
const ctx = canvas.getContext("2d");
ctx.drawImage(image, 0, 0);
document.body.append(canvas);
```

**Reply by calixteman** [reply]
The bug has been reported here:
https://issues.chromium.org/issues/378869810

**Reply by Snuffleupagus** [reply]
Given that there's apparently a few different `ImageDecoder` bugs specifically in Chrome, should we perhaps (for now) by default set the API-option `isImageDecoderSupported = false` in Chrome (and simply get rid of the `isChrome` option)?
